### PR TITLE
NetSim: Limit router count

### PR DESCRIPTION
**Pull request by islemaster**
Limit the number of routers that can be created in a class section.

The maximum number of routers that may be created is the minimum of:
- A global router limit, currently 20, stored as `CDO.netsim_max_routers`
- The maximum number of addressable routers for the level, if applicable.  The number of addressable routers is determined by the bit-width of the second-to-last part of the address format.

Some examples:
- In a level with address format "4" (a single, four-bit part e.g. "0011") there is no router addressing, so we use the global limit of 20 routers.
- In a level with address format "4.4" (two four-bit parts e.g. "0110 0011") we can provide addresses for up to 16 routers.  Since this is lower than the global limit of 20, we allow up to 16 routers.
- In a level with address format "8.8.8.8" we can address up to 256 routers.  Since this is higher than the global limit of 20, we allow up to 20 routers.

When the limit is reached, the "Add Router" button is hidden from the lobby so that no more routers may be created via the client UI.

The server only does partial enforcement of this policy; it has limited knowledge of the level configuration so for now it simply enforces the global router limit of 20.

Finishes [Pivotal #100996328](https://www.pivotaltracker.com/story/show/100996328).

### CODE REVIEW

**Comment by islemaster** [comment]
@sfilman How do you feel about this work getting checked in before we fix the router numbering (which is my next work item)?  Although this change limits the number of routers it does not yet prevent gaps in that numbering, and router numbers still wrap at addressable space so collisions are still possible.  I _think_ it's strictly better than what's in production, but maybe not?

@Hamms I think we can at least review this change in isolation.  Will you review?

File: apps/src/netsim/NetSimRemoteNodeSelectionPanel.js [hunk 1]
```diff
@@ -337,3 +338,20 @@ NetSimRemoteNodeSelectionPanel.prototype.canCurrentUserResetShard = function ()
...
+NetSimRemoteNodeSelectionPanel.prototype.canAddRouter = function () {
+  var levelConfig = NetSimGlobals.getLevelConfig();
+  if (this.hasOutgoingRequest() || !levelConfig.showAddRouterButton) {
+    return false;
+  }
+
+  var routerLimit = NetSimRouterNode.getMaximumRoutersPerShard();
+  var routerCount = this.nodesOnShard_.filter(function (node) {
+        return NodeType.ROUTER === node.getNodeType();
+      }).length;
```
Comment (Hamms) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.js]: is this the indentation format we're using?

File: apps/test/netsim/NetSimRemoteNodeSelectionPanel.js [hunk 2]
```diff
@@ -5,25 +5,162 @@ var NetSimTestUtils = require('../util/netsimTestUtils');
...
+
+      it ("true if current client count is at/beyond the global router limit", function () {
+        var panel = panelWithNodes(makeClients(MAX_ROUTERS));
+        assert(panel.canAddRouter());
+      });
+
+    });
+
+    // The two-bit router part imposes a limit of four addressable routers.
+    describe ("with address format '2.8'", function () {
```
Comment (Hamms) [inline apps/test/netsim/NetSimRemoteNodeSelectionPanel.js]: should this be `"with address format '2.2'"`? Likewise the one below. If not, I'm confused about something.

File: shared/test/test_net_sim_api.rb [hunk 3]
```diff
@@ -31,7 +32,7 @@ def setup
   def test_create_read_update_delete
     # Verify that the CREATE response body and READ response bodies
     # both return the correct record values
-    record_create_response = create_record({name: 'alice', age: 7, male: false})
+    record_create_response = create_record({type: NODE_TYPES[:client], name: 'alice', age: 7, male: false})
```
Comment (Hamms) [inline shared/test/test_net_sim_api.rb]: This indicates to me that we're at a point where we need to start caring a little more about our test data, or at the very least being a bit more explicit about the fact that these records are getting created on the nodes table and therefore have certain requirements.

That, or creating them on another table.

File: apps/test/netsim/NetSimRemoteNodeSelectionPanel.js (hunk 2 above)
Comment (islemaster) [inline apps/test/netsim/NetSimRemoteNodeSelectionPanel.js]: Number of addressable routers is determined by the second-to-last address part (2, in this case); I was hoping to make a point about how the other parts are unimportant for this assertion.

File: shared/test/test_net_sim_api.rb [hunk 4]
```diff
@@ -571,6 +572,114 @@ def test_parse_ids_from_query_string
...
+  def test_do_not_limit_shard_clients_to_max_routers
+    CDO.netsim_max_routers.times do
+      create_node(type: NODE_TYPES[:client])
+      assert_equal 201, @net_sim_api.last_response.status
+    end
+    assert_equal(CDO.netsim_max_routers,
+                 read_records(TABLE_NAMES[:node]).count,
+                 "Didn't create #{CDO.netsim_max_routers} nodes")
+
+    # We want the 21st node to fail
```
Comment (Hamms) [inline shared/test/test_net_sim_api.rb]: nit: `s/fail/succeed`

**Comment by Hamms** [comment]
Couple of tiny nits, but otherwise looks good!

**Comment by islemaster** [comment]
Comments addressed.  Waiting on Travis.

**Pull request merged**